ark/utils/spatial_analysis_utils.py

```diff
@@ -435,60 +437,118 @@ def compute_neighbor_counts(current_fov_neighborhood_data, dist_matrix, distlim,
 
-    return counts, freqs.T
+    for n in range(min_k, max_k + 1):
+        cluster_fit = KMeans(n_clusters=n).fit(neighbor_mat_data)
+        cluster_stats.loc[n] = cluster_fit.inertia_
 
+    return cluster_stats
 
-def compute_kmeans_cluster_metric(neighbor_mat_data, max_k=10):
-    """For a given neighborhood matrix, cluster and compute metric scores using k-means clustering.
 
-    Currently only supporting silhouette score as a cluster metric.
+def compute_kmeans_silhouette(neighbor_mat_data, min_k=2, max_k=10, subsample=None):
+    """For a given neighborhood matrix, cluster and compute Silhouette score using k-means
+       from the range of k=min_k to max_k
 
     Args:
         neighbor_mat_data (pandas.DataFrame):
             neighborhood matrix data with only the desired fovs
+        min_k (int):
+            the minimum k we want to generate cluster statistics for, must be at least 2
         max_k (int):
             the maximum k we want to generate cluster statistics for, must be at least 2
+        subsample (int):
+            the number of cells that will be sampled from each neighborhood cluster for
+            calculating Silhouette score
+            If None, all cells will be used
 
     Returns:
         xarray.DataArray:
-            contains a single dimension, cluster_num, which determines the metric score
+            contains a single dimension, cluster_num, which indicates the Silhouette score
             when cluster_num was set as k for k-means clustering
     """
 
     # create array we can store the results of each k for clustering
-    coords = [np.arange(2, max_k + 1)]
+    coords = [np.arange(min_k, max_k + 1)]
     dims = ["cluster_num"]
     stats_raw_data = np.zeros(max_k - 1)
     cluster_stats = xr.DataArray(stats_raw_data, coords=coords, dims=dims)
 
-    for n in range(2, max_k + 1):
+    for n in range(min_k, max_k + 1):
         cluster_fit = KMeans(n_clusters=n).fit(neighbor_mat_data)
         cluster_labels = cluster_fit.labels_
-        cluster_score = sklearn.metrics.silhouette_score(neighbor_mat_data, cluster_labels,
-                                                         metric='euclidean')
+        neighbor_mat_data["cluster"] = cluster_labels
+
+        if subsample is None:
+            cluster_score = sklearn.metrics.silhouette_score(neighbor_mat_data, cluster_labels,
+                                                             metric="euclidean")
+        else:
+            # Subsample each cluster
+            sub_dat = neighbor_mat_data.groupby("cluster").apply(
+                lambda x: x.sample(subsample, replace=len(x) < subsample)).reset_index(drop=True)
+            cluster_score = sklearn.metrics.silhouette_score(sub_dat.drop("cluster", axis=1),
+                                                             sub_dat["cluster"],
+                                                             metric="euclidean")
```

Comment:
I think this also fixes a bug where the `'cluster'` column wasn't dropped from `neighbor_mat_data` when passing it to `silhouette_score` in the `subsample is None` case.

Reply:
I don't think the original version of the code has that bug, since the `cluster` column is not dropped from `neighbor_mat_data` if `subsample is None`.

I made these suggested changes anyways.